webapp/src/app.tsx

```diff
@@ -1304,27 +1305,38 @@ export class ProjectView
         if (checkAsync)
             return checkAsync.then(() => this.openHome());
 
-        let doSync = false
+        let p = Promise.resolve();
         // check our multi-tab session
         if (workspace.isHeadersSessionOutdated()) {
              // reload header before loading
             pxt.log(`multi-tab sync before load`)
-            doSync = true;
-        }
-        // check our cloud sync
-        const RESYNC_TIME_SEC = 5 * 60 // 5 minutes
-        const headerCloudSyncOutdated = auth.loggedInSync()
-            && Util.nowSeconds() - workspace.getHeaderLastCloudSync(h) > RESYNC_TIME_SEC
-        if (headerCloudSyncOutdated) {
-            pxt.log(`cloud sync before load`)
-            doSync = true;
-        }
-
-        let p = Promise.resolve();
-        if (doSync) {
             p = p.then(() => workspace.syncAsync().then(() => { }))
         }
 
+        // Try a quick cloud fetch. If it doesn't complete within X second(s),
+        // continue on.
+        const timeout = Util.timeout(1500/*1.5 seconds*/)
+        p = p.then(() => {
+            return Promise.any([
+                timeout,
+                // start a partial cloud sync
+                cloud.syncAsync([h])
+                    .then((changes) => {
+                        if (changes.length) {
+                            if (timeout.isResolved()) {
+                                // We are too late; the editor has already been loaded.
+                                // Call the onChanges handler to update the editor.
+                                cloud.onChangesSynced(changes)
```

Comment:
Let's call this project A. Let's assume the connection is slow and the user had the time to leave the editor again and load another project B, are we overriding A sources with B?

Reply:
No, the closure references the "h" that belongs to the particular invocation of loadHeaderAsync that created it. Otherwise there be all sorts of race conditions in async-await code.